### PR TITLE
H2 Settings Acknowledgments do not need to block progress

### LINKED ISSUE #6549: ember server stuck when using an ember client to implement an AuthMiddleware
As promised in **Discord**, I managed to create an MCVE for this issue.

The gist of the problem is here:

```scala
object Service extends Http4sDsl[IO] with Http4sClientDsl[IO] {
  val healthCheck = HttpRoutes.of[IO] {
    case GET -> Root / "healthcheck" =>
      Ok("OK")
  }

  private final case class AuthUser(data: String)

  private def authMiddleware(client: Client[IO], logger: Logger[IO], authUri: Uri): AuthMiddleware[IO, AuthUser] =
    AuthMiddleware(
      authUser = Kleisli { req =>
        req
          .headers
          .get[headers.Authorization]
          .toRight[Throwable](left = new NoSuchElementException("Auth header missing"))
          .flatTraverse { authHeader =>
            val authRequest =
              GET(
                authUri,
                authHeader,
                headers.Accept(MediaType.application.json)
              )

            logger.info(s"Auth middleware - request: ${authRequest}") >>
            client.expect[String](authRequest).flatMap { response =>
              logger.info(s"Auth middleware - response: ${response}").as(AuthUser(response))
            }.attempt
          }
      },
      onFailure = org.http4s.AuthedRoutes.of[Throwable, IO] {
        case _ as ex =>
          InternalServerError(ex.getMessage)
      }
    )

  def appRoutes(client: Client[IO], logger: Logger[IO], authUri: Uri): HttpRoutes[IO] =
    authMiddleware(client, logger, authUri)(
      AuthedRoutes.of[AuthUser, IO] {
        case GET -> Root / "test" as user =>
          Ok(user.data)
      }
    )
}
```

Using **curl**, if I call the `/healthcheck` endpoint, then it works and returns the `Ok` response.
But, if I call the `/api/test` endpoint, then the petition gets stuck and never finishes.

```bash
$ curl http://localhost:8080/healthcheck # OK
$ curl -H "Authorization: Bearer foo" http://localhost:8080/api/test # <stuck>
```

These are the logs:

> 18:36:36.098 [io-compute-8] INFO http4s-error - Auth middleware - request: Request(method=GET, uri=https://postman-echo.com/get, httpVersion=HTTP/1.1, headers=Headers(Authorization: <REDACTED>, Accept: application/json))
> 18:36:36.951 [io-compute-11] DEBUG http4s-error - postman-echo.com:443 Read - Settings(SettingsMaxConcurrentStreams(128), SettingsInitialWindowSize(65536), SettingsMaxFrameSize(16777215))
> 18:36:36.959 [io-compute-1] DEBUG http4s-error - postman-echo.com:443 Write - Settings(SettingsEnablePush(false))
> 18:36:36.959 [io-compute-11] DEBUG http4s-error - postman-echo.com:443 Read - WindowUpdate(0,2147418112)
> 18:36:36.962 [io-compute-1] DEBUG http4s-error - postman-echo.com:443 Write - Settings.Ack
> 18:36:36.964 [io-compute-11] DEBUG http4s-error - ReadLoop has received that is should kill org.http4s.ember.core.h2.H2Connection$KillWithoutMessage
>18:36:36.969 [io-compute-10] DEBUG http4s-error - postman-echo.com:443 Write - GoAway(identifier=0, lastStreamId=0, errorCode=FlowControlError, additionalDebugData=None)

**Note**: Not only did the petition get stuck, but the whole app froze, cancelling the petition did NOT unfreeze it, not even `CTRL+C` or `kill` signals did; I had to use `kill -9` to finish the program.

-----

Please let me know if you need more info, or if I can do anything to help :)

_Hopefully, this is just me doing something stupid_.

----

You can run the complete MCVE using **scala-cli**.

```scala
//> using scala "2.13.8"
//> using lib "org.typelevel::cats-core:2.8.0"
//> using lib "org.typelevel::cats-effect:3.3.14"
//> using lib "org.typelevel::log4cats-slf4j:2.3.2"
//> using lib "ch.qos.logback:logback-classic:1.2.11"
//> using lib "co.fs2::fs2-core:3.2.10"
//> using lib "org.http4s::http4s-dsl:0.23.13"
//> using lib "org.http4s::http4s-ember-client:0.23.13"
//> using lib "org.http4s::http4s-ember-server:0.23.13"

import cats.data.Kleisli
import cats.effect.{IO, IOApp, Resource}
import cats.syntax.all._
import com.comcast.ip4s._
import org.http4s._
import org.http4s.client.Client
import org.http4s.client.dsl.Http4sClientDsl
import org.http4s.dsl.Http4sDsl
import org.http4s.ember.client.EmberClientBuilder
import org.http4s.ember.server.EmberServerBuilder
import org.http4s.server.{AuthMiddleware, Server, Router}
import org.typelevel.log4cats.Logger
import org.typelevel.log4cats.slf4j.Slf4jLogger

import scala.concurrent.duration._

object Main extends IOApp.Simple {
  override final val run: IO[Unit] =
    Slf4jLogger.fromName[IO]("http4s-error").flatMap { logger =>
      logger.info("Starting the server") >>
      HttpServer.server(
        logger,
        authUri = Uri.unsafeFromString("https://postman-echo.com/get"),
      ).evalTap(server =>
        logger.info(s"Server has started at ${server.address}")
      ).useForever.void
    }
}

object HttpServer {
  private def buildApp(client: Client[IO], logger: Logger[IO], authUri: Uri): HttpApp[IO] =
    Router(
      "/" -> Service.healthCheck,
      "/api" -> Service.appRoutes(client, logger, authUri)
    ).orNotFound

  private def buildClient(logger: Logger[IO]): Resource[IO, Client[IO]] =
    EmberClientBuilder
      .default[IO]
      .withLogger(logger)
      .withTimeout(2.seconds)
      .withChunkSize(512)
      .withHttp2
      .build

  private def buildServer(app: HttpApp[IO], logger: Logger[IO]): Resource[IO, Server] =
    EmberServerBuilder
      .default[IO]
      .withHost(host"0.0.0.0")
      .withPort(port"8080")
      .withHttpApp(app)
      .withLogger(logger)
      .withHttp2
      .withErrorHandler {
        case error =>
          logger.error(error)("Unexpected error:").as(Response(status = Status.InternalServerError))
      }.build

  def server(logger: Logger[IO], authUri: Uri): Resource[IO, Server] =
    buildClient(logger).flatMap { client =>
      buildServer(
        app = buildApp(client, logger, authUri),
        logger
      )
    }
}

object Service extends Http4sDsl[IO] with Http4sClientDsl[IO] {
  val healthCheck = HttpRoutes.of[IO] {
    case GET -> Root / "healthcheck" =>
      Ok("OK")
  }

  private final case class AuthUser(data: String)

  private def authMiddleware(client: Client[IO], logger: Logger[IO], authUri: Uri): AuthMiddleware[IO, AuthUser] =
    AuthMiddleware(
      authUser = Kleisli { req =>
        req
          .headers
          .get[headers.Authorization]
          .toRight[Throwable](left = new NoSuchElementException("Auth header missing"))
          .flatTraverse { authHeader =>
            val authRequest =
              GET(
                authUri,
                authHeader,
                headers.Accept(MediaType.application.json)
              )

            logger.info(s"Auth middleware - request: ${authRequest}") >>
            client.expect[String](authRequest).flatMap { response =>
              logger.info(s"Auth middleware - response: ${response}").as(AuthUser(response))
            }.attempt
          }
      },
      onFailure = org.http4s.AuthedRoutes.of[Throwable, IO] {
        case _ as ex =>
          InternalServerError(ex.getMessage)
      }
    )

  def appRoutes(client: Client[IO], logger: Logger[IO], authUri: Uri): HttpRoutes[IO] =
    authMiddleware(client, logger, authUri)(
      AuthedRoutes.of[AuthUser, IO] {
        case GET -> Root / "test" as user =>
          Ok(user.data)
      }
    )
}
```

### PR DESCRIPTION
**Pull request by ChristopherDavenport**
Fixes #6549 

TLDR: Settings Acknowledgements waiting causes a bug on some very present implementations H2.

Long Story: 

Original Framing 
```
DEBUG - postman-echo.com:443 Write - Settings()
DEBUG - postman-echo.com:443 Read - Settings(SettingsMaxConcurrentStreams(128), SettingsInitialWindowSize(65536), SettingsMaxFrameSize(16777215))
DEBUG - postman-echo.com:443 Read - WindowUpdate(0,2147418112)
DEBUG - postman-echo.com:443 Write - Settings.Ack
DEBUG - postman-echo.com:443 Write - GoAway(identifier=0, lastStreamId=0, errorCode=FlowControlError, additionalDebugData=None)
```

As you see, the write for setting proceeds first. Ok, so now for the words in the spec.

```
The values in the SETTINGS frame MUST be processed in the order they
   appear, with no other frame processing between values.  Unsupported
   parameters MUST be ignored.  Once all values have been processed, the
   recipient MUST immediately emit a SETTINGS frame with the ACK flag
   set. 
   ```
   
   So the first thing that the other server should do is to process the settings and Ack. So in the current implementation we waited for this. However we receive a GoAway early because the other server is a jerk.
   
 So after debugging I checked the same framing with nghttp
 
 ```
 nghttp -v https://postman-echo.com/get
[  0.125] Connected
The negotiated protocol: h2
[  0.421] recv SETTINGS frame <length=18, flags=0x00, stream_id=0>
          (niv=3)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):128]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65536]
          [SETTINGS_MAX_FRAME_SIZE(0x05):16777215]
[  0.421] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=0>
          (window_size_increment=2147418112)
[  0.421] send SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
[  0.421] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.421] send PRIORITY frame <length=5, flags=0x00, stream_id=3>
          (dep_stream_id=0, weight=201, exclusive=0)
[  0.421] send PRIORITY frame <length=5, flags=0x00, stream_id=5>
          (dep_stream_id=0, weight=101, exclusive=0)
[  0.421] send PRIORITY frame <length=5, flags=0x00, stream_id=7>
          (dep_stream_id=0, weight=1, exclusive=0)
[  0.421] send PRIORITY frame <length=5, flags=0x00, stream_id=9>
          (dep_stream_id=7, weight=1, exclusive=0)
[  0.421] send PRIORITY frame <length=5, flags=0x00, stream_id=11>
          (dep_stream_id=3, weight=1, exclusive=0)
[  0.421] send HEADERS frame <length=43, flags=0x25, stream_id=13>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /get
          :scheme: https
          :authority: postman-echo.com
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.40.0
[  0.520] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.524] recv (stream_id=13) :status: 200
[  0.524] recv (stream_id=13) date: Thu, 21 Jul 2022 03:00:07 GMT
[  0.524] recv (stream_id=13) content-type: application/json; charset=utf-8
[  0.525] recv (stream_id=13) content-length: 279
[  0.525] recv (stream_id=13) etag: W/"117-ui6vXK51fkAZ5qjdNV072ppgReA"
[  0.525] recv (stream_id=13) vary: Accept-Encoding
[  0.525] recv (stream_id=13) set-cookie: sails.sid=s%3A_kz4MkXfTS-CirMQBYDL-9Qvx693uiU4.B96HvMa1FyivP%2BhDE8yAwHVe6e7yApj3RYLiX5r0NjY; Path=/; HttpOnly
[  0.525] recv HEADERS frame <length=203, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
          ; First response header
{"args":{},"headers":{"x-forwarded-proto":"https","x-forwarded-port":"443","host":"postman-echo.com","x-amzn-trace-id":"Root=1-62d8c137-00bf05f57d156fe427885316","accept":"*/*","accept-encoding":"gzip, deflate","user-agent":"nghttp2/1.40.0"},"url":"https://postman-echo.com/get"[  0.525] recv DATA frame <length=278, flags=0x00, stream_id=13>
}[  0.525] recv DATA frame <length=1, flags=0x00, stream_id=13>
[  0.525] recv DATA frame <length=0, flags=0x01, stream_id=13>
          ; END_STREAM
[  0.525] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=NO_ERROR(0x00), opaque_data(0)=[])
```

The critical difference here is that they send the Headers frame before they receive the acknowledgement. I switched it and we consistently get the right response back from the other servers.

It seems that whatever is implementing this server as well as several others requires that clients send the headers before it will acknowledge the connection settings. and otherwise gives an erroneous FlowControlError (which should only be of bad window updates and we clearly are using the default frame size. Note all servers with this bad setup give the exact same non-default settings, leading me to believe this is 1 specific implementation.

```
DEBUG - postman-echo.com:443 Read - Settings(SettingsMaxConcurrentStreams(128), SettingsInitialWindowSize(65536), SettingsMaxFrameSize(16777215))
DEBUG - postman-echo.com:443 Read - WindowUpdate(0,2147418112)
DEBUG - postman-echo.com:443 Write - Settings()
DEBUG - postman-echo.com:443 Write - Settings.Ack
DEBUG - postman-echo.com:443 Write - Ping.Ack
DEBUG - postman-echo.com:443 Write - Headers(identifier=1, dependency=None, endStream=false, endHeaders=true, headerBlock=ByteVector(20 bytes, 0x828744836262a7418bace84d23a9629273ae43d3), padding=None)
DEBUG - postman-echo.com:443 Write - Data(identifier=1, data=ByteVector(empty), pad=None, endStream=true)
DEBUG - postman-echo.com:443 Read - Settings.Ack
DEBUG - postman-echo.com:443 Read - Headers(identifier=1, dependency=None, endStream=false, endHeaders=true, headerBlock=ByteVector(204 bytes, 0x886196df3dbf4a082a65b685040109403371a6ae084a62d1bf5f961d75d0620d263d4c7441eafb50938ec415305a99567b5c0332333000832a47379de4c7f25716ee8fdaeaf137efbeec6c26d8fc3e903d870feeebbb1afe7f0004766172798b84842d695b05443c86aa6f00874152b10e7ea62fd740cd421741a4810ab30ede5b01e33ea8c9ab17ed3161ebab5b8e59b3f11eeca7258d81d5dbe5dd2f4b6d21c75aa2ba5cb92acf768fcc1cb4e0171c9dbf3ded2f3671d99639ff6cc00fb5358d33c0c7da98d29af55547af), padding=None)
DEBUG - postman-echo.com:443 Read - Data(identifier=1, data=ByteVector(229 bytes, 0x7b2261726773223a7b7d2c2268656164657273223a7b22782d666f727761726465642d70726f746f223a226874747073222c22782d666f727761726465642d706f7274223a22343433222c22686f7374223a22706f73746d616e2d6563686f2e636f6d222c22782d616d7a6e2d74726163652d6964223a22526f6f743d312d36326438636239362d316236303835316235646432633630653038356130396263222c227472616e736665722d656e636f64696e67223a226368756e6b6564227d2c2275726c223a2268747470733a2f2f706f73746d616e2d6563686f2e636f6d2f67657422), pad=None, endStream=false)
DEBUG - postman-echo.com:443 Read - Data(identifier=1, data=ByteVector(1 bytes, 0x7d), pad=None, endStream=false)
DEBUG - postman-echo.com:443 Read - Data(identifier=1, data=ByteVector(empty), pad=None, endStream=true)
```

As a result. I think this brings our implementation into line with others, and working. But isn't actually our server doing anything wrong per the spec. 

However not waiting lets us go faster, which is a good thing!

### CODE REVIEW

**Review by armanbilge — APPROVED** [review]
🎉

**Review by isomarcte — APPROVED** [review]
